akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+    * "Real world" calls of [[invokeWithFeedback()]] always return failed promises for `Completed` state
+    */
+  private class ConcurrentAsyncCallback[T](handler: T ⇒ Unit) extends AsyncCallback[T] {
+
+    sealed trait State
+    //waiting for materialization completion
+    private case class Pending(pendingEvents: List[Event]) extends State
+    //GraphStage sending all events to stream
+    private case object Initializing extends State
+    //stream is initialized and so no threads can just send events without any synchronization overhead
+    private case object Initialized extends State
+    //stage has been shut down, either regularly or it failed
+    private case object Completed extends State
+    //Event with feedback promise
+    private case class Event(e: T, handlingPromise: OptionVal[Promise[Done]])
+
+    val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+    private[this] val currentState = new AtomicReference[State](Pending(List.empty[Event]))
+
+    //is called from stream - thread pool is managed by Akka
+    @tailrec
+    final private[stage] def onStart(): Unit = {
+      (currentState.getAndSet(Initializing): @unchecked) match {
+        case Pending(l) ⇒ l.reverse.foreach(ack ⇒ {
+          onAsyncInput(ack.e)
+        })
+      }
+      if (!currentState.compareAndSet(Initializing, Initialized)) {
+        (currentState.get: @unchecked) match {
+          case Pending(_) ⇒ onStart()
+          case Completed  ⇒ () //wonder if this is possible
+        }
+      }
+    }
+
+    //is called from stream - thread pool is managed by Akka
+    final private[stage] def onStop(): Unit = {
+      currentState.set(Completed)
+      concurrentSet.asScala.foreach(_.tryFailure(new StreamDetachedException()))
+      concurrentSet.clear()
+    }
+
+    private def onAsyncInput(event: T) = interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+
+    private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
+      onAsyncInput(event)
+      currentState.get() match {
+        case Completed ⇒ failPromiseOnComplete(promise)
+        case _         ⇒ promise
+      }
+    }
+
+    private def createFeedback(): Promise[Done] = {
+      val promise: Promise[Done] = Promise[Done]()
+      concurrentSet.add(promise)
+      promise.future.andThen { case _ ⇒ concurrentSet.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      promise
+    }
+
+    //external call
+    override def invokeWithFeedback(event: T): Future[Done] = invokeWithPromise(event).result.future
+
+    private def invokeWithPromise(event: T): TailRec[Promise[Done]] = {
+      val promise: Promise[Done] = createFeedback
+
+      currentState.get() match {
+        //not started yet
+        case list @ Pending(_) ⇒ tailcall(onPending(list, promise, event))
+        //started - can just send message to stream
+        case Initialized       ⇒ done(sendEvent(event, promise))
```

Comment:
Do we have to use these `TailCall` things? Looks like they allocate things, such as this `Done()`

Reply:
Not only the actual `TailCall` objects, but `by-name`-functions everywhere they actually tailcall. I think another option would be preferable. Does just calling into methods right away cause a too deep stack or what is the reason for using the tailcall API? (I tried just doing that on top of the PR it and it seems to work)

Reply:
The idea here was to do trampoline. As I remember it was not possible to do tailrec

Reply:
let me revise this part (actually it should be two methods with such approach here)